feat: webhook workers

steebchen opened the pull request:
# Description

<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

fix #347 

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

## What's Changed

## TODO

- [x] webhook request check (mark step as failed if webhook returns non-200)
- [x] basic healthcheck (call API route before starting)
- [x] advanced healtcheck (ping continuously and take down worker if more than 3 failed healthchecks)
- [x] change action names to get from healthcheck instead of query
- [x] return response only once stuff is processed
- [x] do not require workflow IDs to be specified in registerWorkflow (or UI), do that automatically instead
- [x] think about: changing workflows (with their actions) or assigning new workflows to webhook workers, and killing existing ones/re-creating (internal note: check both worfklows and actions (via workflow hash?) and re-create if needed)
- [x] !!! the healthcheck needs to automatically update/register the workflow (it doesn't seem to be registering it the first time?)
	- [x] Go client
		- [x] requires refactor of .On->.RegisterWorkflow methods and signature
	- [x] TS client
- [x] TS client: check for signature as middleware
- [x] !!! verify webhook secret in health endpoint to make sure it is set correctly before sending any events
- [x] !! cleanup hardcoded env vars (token & tlsStrategy)
- [x] !! do not re-create token every time
- [x] investigate bug where no healthcheck requests are sent (might be invalid now, not sure)
- [x] encrypt token
- [x] rewrite queries to sqlc x)
- [ ] test services
- [ ] test register action
- [x] delete webhook workers
	- [x] (API & UI)
	- [x] stop currently running workers
- [x] add webhook worker name to DB and UI
- [ ] optional: vercel streaming or waitUntil

steebchen committed:
feat: webhook workers

steebchen committed:
fix cache?

steebchen committed:
spawn workers from webhook workers

steebchen committed:
WIP webhooks api

steebchen committed:
continuously spwan webhook workers

steebchen committed:
fix contracts

steebchen committed:
implement webhook worker create

steebchen committed:
adapt tests

steebchen committed:
temp

steebchen committed:
temp rename to test

steebchen committed:
Revert "temp"

This reverts commit a7a1c72cc72e99e2067d6c4328afba450edfe5b0.

steebchen committed:
migration

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
fix leaking goroutine

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
implement webhook worker relations to workflows

steebchen committed:
use correct webhook worker data

steebchen committed:
upsert webhook worker

steebchen committed:
migration

steebchen committed:
allow setting secret, generate if not set

steebchen committed:
render webhook workers properly

steebchen committed:
ui: webhook worker creation popup

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
adapt webhook-workers.yaml

steebchen committed:
upsert webhook worker relations

steebchen committed:
migrations

steebchen committed:
use json tags for action type for unified api

steebchen committed:
upsert webhook worker with id OR name

steebchen committed:
Revert "upsert webhook worker with id OR name"

This reverts commit 96a27b77d26942a793b5118389a771854cd66f15.

steebchen committed:
refactor worker

steebchen committed:
upsert webhook worker with id OR name

steebchen committed:
add webhook handler options struct

steebchen committed:
add webhook healthchecks, remove actions query

steebchen committed:
ensure sending response after event is sent

steebchen committed:
refactor test

steebchen committed:
refactor e2e test with prisma

steebchen committed:
more webhook test refactors

steebchen committed:
mark action as failed on webhook failure & add test

steebchen committed:
add default timeout of 10 minutes for webhook request

steebchen committed:
unskip

steebchen committed:
fix e2e test

steebchen committed:
prisma workaround

steebchen committed:
fix port

steebchen committed:
add "advanced" healthcheck

steebchen committed:
automatically register workflows on creation

steebchen committed:
adapt logs

steebchen committed:
restart on workflow or step changes

steebchen committed:
set webhook worker name

steebchen committed:
remove security from webhook-worker

steebchen committed:
remove hardcoded env setup

steebchen committed:
generate tenant token on webhook ctrl start

steebchen committed:
temp test set server

steebchen committed:
temp set tls and enc

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
Revert "temp set tls and enc"

This reverts commit 9c84024633dbb8fa64496b8f09472515dace58f7.

steebchen committed:
Revert "temp test set server"

This reverts commit f69f23e917716daed63366a7e4997f97e232ff15.

steebchen committed:
Revert "remove hardcoded env setup"

This reverts commit 682c737a

steebchen committed:
temp set token via env var

steebchen committed:
remove hardcoded env setup

This reverts commit b16789152cfd0b0ad2869e0e68497d70c3c30ad8.

steebchen committed:
tls strategy

steebchen committed:
Revert "tls strategy"

This reverts commit de79fe253522cca2f73006ca15295e410c64a918.

steebchen committed:
refactor token and healthcheck logic

steebchen committed:
fix loop

steebchen committed:
refactor webhook worker var naming

steebchen committed:
use .WithToken() instead of env var

steebchen committed:
add logs

steebchen committed:
add log to StartStepRun

steebchen committed:
log step runs

steebchen committed:
attempt to fix client WithToken

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
undo stuff & adapt logs

steebchen committed:
remove workflow selector in create webhook worker

steebchen committed:
release ticker

steebchen committed:
change wf debug log

steebchen committed:
change healthcheck logic, expect signature, unify request

steebchen committed:
use correct log

steebchen committed:
fix test mock healthcheck

steebchen committed:
fix webhook handler get request

steebchen committed:
save token

steebchen committed:
merge main, move stuff

steebchen committed:
fixes

steebchen committed:
fix token tests

steebchen committed:
migration

steebchen committed:
use -y for npx in generate clients

steebchen committed:
attempt to fix integration test

steebchen committed:
wip

steebchen committed:
webhook worker name (wip)

steebchen committed:
webhook worker name (continuation)

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

# Conflicts:
#	api/v1/server/oas/gen/openapi.gen.go

steebchen committed:
migration

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
Merge remote-tracking branch 'origin/main' into webhook-workers

steebchen committed:
adapt

steebchen committed:
use webhook worker id for worker name

steebchen committed:
feat(worker): add RegisterWorkflow method

steebchen committed:
automatically register workflows

steebchen committed:
generate

steebchen committed:
encrypt webhook worker token

steebchen committed:
rewrite webhook workers to sqlc

steebchen committed:
delete webhook worker API route & UI

steebchen committed:
add ww deleted, remove workers on deletion

steebchen committed:
Merge branch 'main' into webhook-workers

steebchen committed:
fix healthcheck response after register workflows

abelanger5 commented on api-contracts/openapi/components/schemas/webhook_worker.yaml:
```diff
@@ -0,0 +1,68 @@
+WebhookWorker:
+  x-resources: ["tenant", "event"]
```
Don't need `x-resources` defined for components, only paths.

abelanger5 commented on api-contracts/openapi/components/schemas/webhook_worker.yaml:
```diff
@@ -0,0 +1,68 @@
+    url:
+      type: string
+      description: The webhook url.
+    secret:
```
We probably don't want to pass this secret key with every webhook worker response. We probably only want to send it after the initial create endpoint is called with a separate component definition.

abelanger5 commented on api-contracts/openapi/components/schemas/webhook_worker.yaml:
```diff
@@ -0,0 +1,68 @@
+      description: The workflow IDs or names to register for this webhook worker. If not provided, workflows will be automatically detected.
+      items:
+        type: string
+    secret:
```
Is this used/does it make sense for the user to set the secret value?

steebchen replied:
I added it so we can set it in our tests, if we really don't want to allow users to set it ever I could workaround it by editing the DB in the tests, but it seemed like to much trouble. So technically a user can set this via the API, but it's not exposed in the UI.

abelanger5 replied:
Makes sense

abelanger5 commented on api-contracts/openapi/components/schemas/webhook_worker.yaml:
```diff
@@ -0,0 +1,68 @@
+
+WebhookWorkerDeleteRequest:
+  properties:
+    id:
```
Instead of the ID being in the request body, we should place it as a URI param to be consistent with the rest of the API.

abelanger5 commented on api-contracts/openapi/openapi.yaml:
```diff
@@ -160,3 +160,9 @@ paths:
     $ref: "./paths/github-app/github-app.yaml#/repos"
   /api/v1/github-app/installations/{gh-installation}/repos/{gh-repo-owner}/{gh-repo-name}/branches:
     $ref: "./paths/github-app/github-app.yaml#/branches"
+  /api/v1/webhook-workers/{tenant}/list:
```
```suggestion
  /api/v1/tenants/{tenant}/webhook-workers:
```

(for consistency)

abelanger5 commented on api-contracts/openapi/openapi.yaml:
```diff
@@ -160,3 +160,9 @@ paths:
     $ref: "./paths/github-app/github-app.yaml#/branches"
+  /api/v1/webhook-workers/{tenant}/list:
+    $ref: "./paths/webhook-worker/webhook-worker.yaml#/list"
+  /api/v1/webhook-workers/{tenant}/create:
```
For consistency, this should be part of the `/api/v1/tenants/{tenant}/webhook-workers` endpoint.

abelanger5 commented on api-contracts/openapi/openapi.yaml:
```diff
@@ -160,3 +160,9 @@ paths:
+    $ref: "./paths/webhook-worker/webhook-worker.yaml#/list"
+  /api/v1/webhook-workers/{tenant}/create:
+    $ref: "./paths/webhook-worker/webhook-worker.yaml#/create"
+  /api/v1/webhook-workers/{tenant}/delete:
```
```suggestion
  /api/v1/webhook-workers/{webhook}:
```

(for consistency)

abelanger5 commented on api-contracts/openapi/paths/webhook-worker/webhook-worker.yaml:
```diff
@@ -0,0 +1,129 @@
+        description: Method not allowed
+    summary: Create a webhook
+delete:
+  post:
```
```suggestion
  delete:
```

abelanger5 commented on api-contracts/openapi/paths/webhook-worker/webhook-worker.yaml:
```diff
@@ -0,0 +1,129 @@
+  post:
+    description: Deletes a webhook
+    operationId: webhook:delete
+    x-resources: ["tenant"]
```
```suggestion
    x-resources: ["tenant", "webhook"]
```

Note that when you add `webhook` to the path param here, you'll want to specify `webhook` as part of the resource hierarchy as well, which will automatically populate the tenant resource.

abelanger5 commented on api/v1/server/run/run.go:
You will need to add a `getter` method to this file when you include `webhook` in the path URI. Something like:

```go
populatorMW.RegisterGetter("webhook", func(config *server.ServerConfig, parentId, id string) (result interface{}, uniqueParentId string, err error) {
		webhook, err := ... // TODO: method
	        if err != nil {
			return nil, "", err
		}
		return webhook, webhook.TenantId, nil
	})
```
	
Note that this getter returns the resource and also a parent id for the resource, which is the parent in the `x-resources` array (which is what defines the resource tree).

abelanger5 commented on examples/webhook/main_e2e_test.go:
💯 awesome to see an e2e test for this

abelanger5 commented on internal/randstr/randstr.go:
```diff
@@ -0,0 +1,44 @@
+	return GenerateRandomString(32)
+}
+
+func MustGenerateRandomString(n int) string {
```
The `encryption` service has a `GenerateRandomBytes` method. It would be great to consolidate using one method or the other. You are free to replace all usage of `GenerateRandomBytes` with this package (would be great to add a simple unit test as well)

steebchen committed:
test service & RegisterAction

abelanger5 commented on internal/services/webhooks/webhooks.go:
```diff
@@ -0,0 +1,317 @@
+			if _, ok := c.registeredWorkerIds[id]; ok {
+				if ww.Deleted {
+					c.sc.Logger.Debug().Msgf("webhook worker %s of tenant %s has been deleted", id, tenantId)
+					err := c.sc.EngineRepository.Worker().DeleteWorkersByName(context.Background(), tenantId, "Webhook_"+id)
```
Why aren't we just using the `DeleteWorker` method for this? Don't we have access to the associated worker id?

steebchen replied:
we don't have access to the worker ID, no. We would need to track it in the DB and while using the worker name does not seem perfect, neither seems tracking this additionaly in the DB because you mentioned that you do not want too much webhook workers logic as this ideally could be done outside of the engine as well

abelanger5 replied:
Ok, that makes sense - but does it even make sense to delete the worker vs mark the worker as inactive? We don't currently delete workers, because it's useful to look back at the run history sometimes.

abelanger5 commented on internal/services/webhooks/webhooks.go:
```diff
@@ -0,0 +1,317 @@
+			return fmt.Errorf("could not get webhook workers: %w", err)
+		}
+
+		for _, ww := range wws {
```
We have a couple of blocking operations in this for loop. Should we consider using an `errgroup` or some other `sync` primitive to run sets of these registrations concurrently?

abelanger5 commented on internal/services/webhooks/webhooks.go:
```diff
@@ -0,0 +1,317 @@
+							c.sc.Logger.Err(err).Msgf("could not update worker")
+						}
+					} else {
+						c.sc.Logger.Printf("webhook worker %s of tenant %s failed one health check, retrying...", id, tenantId)
```
scope this to `Warn` level?

abelanger5 commented on internal/services/webhooks/webhooks.go:
```diff
@@ -0,0 +1,317 @@
+
+					delete(c.registeredWorkerIds, id)
+
+					continue
```
This `continue` seems unnecessary

abelanger5 commented on pkg/repository/prisma/dbsqlc/api_tokens.sql:
```diff
@@ -22,3 +22,21 @@ INSERT INTO "APIToken" (
     @expiresAt::timestamp
 ) RETURNING *;
+
+-- name: UpsertAPIToken :one
```
Where is this used? Not finding it

steebchen replied:
was unused from an old implementation 👍

abelanger5 commented on prisma/schema.prisma:
```diff
@@ -91,6 +91,41 @@ model UserSession {
+  tokenId String?   @db.Uuid
+  token   APIToken? @relation(fields: [tokenId], references: [id], onDelete: Cascade, onUpdate: Cascade)
+
+  // TODO remove tenant ID as it is not needed
```
address TODO?

abelanger5 commented on pkg/worker/webhook_handler.go:
```diff
@@ -0,0 +1,158 @@
+
+		data, err := io.ReadAll(r.Body)
+		if err != nil {
+			panic(err)
```
Should we be panicking in this method?

abelanger5 commented on api/v1/server/handlers/webhook-worker/create.go:
```diff
@@ -0,0 +1,44 @@
+		TenantId:  tenant.ID,
+		Name:      request.Body.Name,
+		URL:       request.Body.Url,
+		Secret:    secret,
```
We should be storing the webhook secret encrypted as well.

abelanger5 commented on api/v1/server/handlers/webhook-worker/create.go:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	var wfs []string
+	if request.Body.Workflows != nil {
```
This is sent through the healthcheck now, right?

steebchen replied:
yes, in theory we can remove this part, not sure if this would be useful in the future or not

abelanger5 replied:
I think it would make sense to remove it if it's unused, otherwise it gets confusing to track where workflows are getting registered

abelanger5 commented on pkg/worker/webhook_handler.go:
```diff
@@ -0,0 +1,158 @@
+
+type HealthCheckResponse struct {
+	Actions   []string `json:"actions"`
+	Workflows []string `json:"workflows"`
```
Shouldn't `Workflows` be a complete workflows object? We'd want to register the correct concurrency settings, timeout settings, etc, right?

steebchen replied:
This just exists to register the actions for the events – the actual workflow is registered separately and thus all settings like concurrency etc should be registered correctly

abelanger5 replied:
Got it, where is the actual workflow registered from?

steebchen replied:
In the respective SDKs in the healthcheck response right now

steebchen committed:
fix minor issues

steebchen committed:
add webhook worker api repo

steebchen committed:
improve ww api routes

steebchen committed:
remove invalid TODO

steebchen committed:
remove continue

steebchen committed:
add WebhookWorkerCreated model to returns secret once

steebchen committed:
feat(random): refactor random string generation

- move method to new random package
- rename GenerateToRandomBytes to GenerateRandomString
- use specific characters for random string
- add tests

steebchen committed:
fix webhook handler response

steebchen committed:
feat(encryption): add String using base64 variants

steebchen committed:
encrypt webhook worker secret

steebchen committed:
update random methods

steebchen committed:
use errgroup for webhook registration

steebchen committed:
fix hatchet-admin token create command

steebchen committed:
Merge branch 'main' into webhook-workers

steebchen committed:
fix loop closure

steebchen committed:
squash migration

steebchen committed:
remove unused workflows slice in WW upsert

steebchen committed:
set worker inactive instead of deleting

steebchen committed:
doc update

steebchen committed:
a bunch more healthcheck related improvements

steebchen committed:
ww test names

steebchen committed:
merge main

abelanger5 commented on api-contracts/openapi/openapi.yaml:
```diff
@@ -138,15 +138,7 @@ paths:
-    $ref: "./paths/webhook-worker/webhook-worker.yaml#/delete"
+  /api/v1/tenants/{tenant}/webhook-workers:
+    $ref: "./paths/webhook-worker/webhook-worker.yaml#/webhookworkers"
+  /api/v1/tenants/webhook-workers/{webhook}:
```
```suggestion
  /api/v1/webhook-workers/{webhook}:
```

(for consistency)

abelanger5 commented on internal/services/webhooks/webhooks.go:
```diff
@@ -76,88 +79,108 @@ func (c *WebhooksController) check() error {
-				c.cleanups = append(c.cleanups, cleanup)
-			}
+		if err = ec.Wait(); err != nil {
+			panic(err) // the error group above is not expected to error
```
I'm still not sure we should panic here, or the entire engine will restart right? Unless there's somewhere you're handling panic/recover.

steebchen replied:
this line will never happen, I could have written _  = ec.Wait() as well

steebchen replied:
I will rewrite this

steebchen replied:
done, it's just using a goroutine now

steebchen committed:
adapt webhook-workers api url

steebchen committed:
webhook healthcheck use regular goroutine

grutt commented on api-contracts/openapi/components/schemas/webhook_worker.yaml:
```diff
@@ -0,0 +1,69 @@
+    - metadata
+    - name
+    - url
+    - secret
```
```suggestion

```

grutt replied:
looks like this is unused?

steebchen replied:
good catch, interesting that this doesn't error as the field doesn't exist

grutt commented on api-contracts/openapi/paths/webhook-worker/webhook-worker.yaml:
```diff
@@ -0,0 +1,123 @@
+          application/json:
+            schema:
+              $ref: "../../components/schemas/_index.yaml#/WebhookWorkerListResponse"
+        description: Successfully changed password
```
```suggestion
        description: The list of webhooks
```

grutt commented on api/v1/server/handlers/webhook-worker/list.go:
```diff
@@ -0,0 +1,33 @@
+func (i *WebhookWorkersService) WebhookList(ctx echo.Context, request gen.WebhookListRequestObject) (gen.WebhookListResponseObject, error) {
+	tenant := ctx.Get("tenant").(*db.TenantModel)
+
+	webhooks, err := i.config.EngineRepository.WebhookWorker().ListActiveWebhookWorkers(context.TODO(), tenant.ID)
```
```suggestion
	webhooks, err := i.config.EngineRepository.WebhookWorker().ListActiveWebhookWorkers(context.Background(), tenant.ID)
```

grutt replied:
unless there's a reason

grutt commented on frontend/app/src/pages/main/tenant-settings/webhooks/index.tsx:
```diff
@@ -0,0 +1,205 @@
+            showDialog={showCreateTokenDialog}
+            setShowDialog={setShowCreateTokenDialog}
+            onSuccess={() => {
+              void listWebhookWorkersQuery.refetch();
```
nit: invalidate the query instead of refetching

steebchen replied:
I do need to refetch the data so that the newly added item shows, no? Or what am I missing?

grutt replied:
```
queryClient.invalidateQueries({
        queryKey: queries.workflowRuns.get(
          tenant.metadata.id,
          workflowRun.metadata.id,
        ).queryKey,
      });

```

Something like this will invalidate all requests with that query and refetch accordingly, i think refetch will just refetch that specific request

grutt replied:
minor thing and not required for this pr

steebchen replied:
if the new response is cached it would not matter though I think, is it not cached atm?

grutt commented on frontend/app/src/router.tsx:
```diff
@@ -209,6 +209,17 @@ const routes: RouteObject[] = [
                   ),
               },
+              {
+                path: '/tenant-settings/webhooks',
```
nit: call the page webhook-workers since at some point we'll likely add webhooks for workflow run events?

steebchen replied:
hmm yeah, though I'm not sure if we want to expose the name Webhook Workers to the users as I was thinking it might be confused with regular Workers. Maybe "Webhook Apps" although that also sounds weird (Inngest uses the term app)

grutt replied:
we can change it later 🫠

steebchen replied:
aka never 😂

grutt commented on internal/services/webhooks/webhooks.go:
```diff
@@ -0,0 +1,342 @@
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
```
less frequent? maybe once per minute?

at some point it would be nice to "invalidate" on code changes

steebchen replied:
good catch I decreased it for testing

steebchen replied:
Whenever it runs this it compares the hash of the actions and workflows, so it kinda invalidates on code changes if that's what you meant

grutt replied:
nice, i meant more for adding actions but i dont think its critical

grutt commented on internal/whrequest/request.go:
```diff
@@ -0,0 +1,59 @@
+
+	httpClient := &http.Client{
+		// use 10 minutes timeout
+		Timeout: time.Second * 600,
```
why 10m?

steebchen replied:
Good point – I chose this randomly because the default http client doesn't have a timeout. I figured workflows which run over http should not really take that long, but it's set completely arbitrary and if it says we should document it

grutt commented on prisma/schema.prisma:
```diff
@@ -90,6 +90,40 @@ model UserSession {
+  updatedAt DateTime @default(now()) @updatedAt
+
+  name       String
+  secret     String
```
check: secret encrypted

grutt replied:
looks good

grutt commented on prisma/schema.prisma:
```diff
@@ -90,6 +90,40 @@ model UserSession {
+  secret     String
+  url        String  @unique
+  // stores the encrypted hatchet auth token
+  tokenValue String?
```
why are we storing this here and also in token?

steebchen replied:
Where in `token` do you mean?

grutt replied:
my fault

grutt reviewed (approved):
good work.
a few minor nits and questions but probably nothing show stopping

steebchen committed:
address review, adapt context
